deployer/src/deployment/build_logs.rs

```diff
@@ -0,0 +1,200 @@
+use super::{BuildLogReceiver, BuildLogSender};
+
+use std::collections::HashMap;
+use std::io;
+use std::sync::{Arc, Weak};
+
+use tokio::sync::{broadcast, RwLock};
+
+const BUFFER_SIZE: usize = 10;
+
+#[derive(Clone)]
+pub struct BuildLogsManager {
+    deployments: Arc<RwLock<HashMap<String, Deployment>>>,
+}
+
+impl BuildLogsManager {
+    pub fn new() -> Self {
+        BuildLogsManager {
+            deployments: Arc::new(RwLock::new(HashMap::new())),
+        }
+    }
+
+    pub async fn for_deployment(&self, name: String) -> BuildLogWriter {
+        let (sender, mut receiver) = broadcast::channel(BUFFER_SIZE);
+
+        let sender = Arc::new(sender);
+        let sender_weak = Arc::downgrade(&sender);
+
+        let logs_so_far = Arc::new(RwLock::new(Vec::new()));
+
+        let logs_so_far_clone = logs_so_far.clone();
+        // This Tokio task is responsible for receiving build log lines and
+        // storing them in the `logs_so_far` vector.
+        let log_store_task_handle = tokio::spawn(async move {
+            while let Ok(line) = receiver.recv().await {
+                logs_so_far_clone.write().await.push(line);
+            }
+        });
+
+        self.deployments.write().await.insert(
+            name,
+            Deployment {
+                sender: sender_weak,
+                logs_so_far: logs_so_far.clone(),
+                log_store_task_handle,
+            },
+        );
+
+        BuildLogWriter {
+            sender,
+            buffer: String::new(),
+        }
+    }
+
+    pub async fn delete_deployment(&self, name: &str) {
+        self.deployments.write().await.remove(name);
+    }
+
+    pub async fn subscribe(&self, name: &str) -> Option<BuildLogReceiver> {
+        self.deployments
+            .read()
+            .await
+            .get(name)
+            .and_then(|deployer| deployer.sender.upgrade())
+            .map(|sender| sender.subscribe())
+    }
+
+    pub async fn get_logs_so_far(&self, name: &str) -> Option<Vec<String>> {
+        if let Some(deployment) = self.deployments.read().await.get(name) {
+            Some(deployment.logs_so_far.read().await.clone())
+        } else {
+            None
+        }
+    }
+}
+
+struct Deployment {
+    /// A weak pointer to the [`BuildLogSender`] stored in a [`BuildLogWriter`].
+    /// It is a weak pointer so that it is possible to identify when a build
+    /// has failed/ended (and thus close a WebSocket connection sending build
+    /// logs). The dropping of `BuildLogWriter` will result in the closing of
+    /// the build log broadcast channel as this weak pointer cannot keep the
+    /// channel open.
+    sender: Weak<BuildLogSender>,
+    logs_so_far: Arc<RwLock<Vec<String>>>,
```

Comment:
Shouldn't we be pushing the logs to the log table rather than keeping them in memory?

Reply:
Sorry @chesedo , do you have any suggestions of how to approach this with regards to your logging implementation? I'm not failure with `tracing` and am struggling to get it work how I want :sweat_smile:

Reply:
When I was working on the tests for #235, it appeared the build logs are automatically caught by tracing (and therefore persistence). We may just need to edit the parsing of these logs in `deploy_layer.rs` a bit :sweat_smile: